[Security Solution][DE] Migrate investigation_fields

yctercero opened the pull request:
## Summary

**TLDR:** SO will support both `string[]` and `{ field_names: string[] }`, but detection engine APIs will only support the object format in 8.11+.

PR seems large but most changes are test changes.

Adds logic to migrate `investigation_fields` from `string[]` to `{ field_names: string[] }` on read. The exception to this is the import route which will transform the legacy investigation field prior to creating the rules so that users can continue to import from 8.10 without issue. A missed backport for 8.10 resulted in this field being typed as `string[]` when intended to be an object.

The logic is conducting the migration at the HTTP level, transforming the field on read. It does also migrate in two other places:
- on import, prior to validation so that users can export from 8.10 and import without issue
- on the client side, there is one hook that fetches a rule, and if the response is a 404, then it is assumed that the rule is deleted and it tries to query for the last alert created by the rule. Since we are not looking to migrate any alert docs, and this is the only scenario in the UI where we are reading from the alert instead of the rule for `investigation_fields`, we do a simple transform on read there. 

| Method | Endpoint                                                             | Migrates SO | Transforms Response | Legacy blocked in request
| ------- | ----------------------------------------------------------------- | ----------- | ----------- | ----------- |
| POST   | /api/detection_engine/rules                               |                         |                                     |  ✅               |
| GET    | /api/detection_engine/rules                                |                         | ✅                                 |               |
| PUT    | /api/detection_engine/rules                                |                         | ✅                                      |  ✅                     |
| PATCH  | /api/detection_engine/rules                              |                        |  ✅                                |  ✅                |
| DELETE | /api/detection_engine/rules                              |                       | ✅                                     |                    |
| POST   | /api/detection_engine/rules/\_bulk_create       |                        |                                      | ✅                 |
| PUT    | /api/detection_engine/rules/\_bulk_update       |                         | ✅                                         | ✅          |
| PATCH  | /api/detection_engine/rules/\_bulk_update      |                       | ✅                                       |  ✅            |
| DELETE | /api/detection_engine/rules/\_bulk_delete       |                      | ✅                                     |                         |
| POST   | /api/detection_engine/rules/\_bulk_delete        |                         | ✅                                   |         |
| POST   | /api/detection_engine/rules/_export                                     |                         | ✅                                   |         |
| GET   | /api/detection_engine/rules/_find                                          |                         | ✅                                   |         |
| POST   | /api/detection_engine/rules/_import                                 |   ✅ (migrates prior to creating the rules)     |            |         |
| POST   | /api/detection_engine/rules/_bulk_action  - export           |         | ✅           |         |
| POST   | /api/detection_engine/rules/_bulk_action  - delete           |         | ✅           |         |
| POST   | /api/detection_engine/rules/_bulk_action  - enable           |         | ✅           |         |
| POST   | /api/detection_engine/rules/_bulk_action  - disable           |         | ✅           |         |
| POST   | /api/detection_engine/rules/_bulk_action  - duplicate           |        |  ✅   |         |
| POST   | /api/detection_engine/rules/_bulk_action  - edit           |        |    ✅        |         |
| POST   | /api/detection_engine/rules/{id}/exceptions                   |         | ✅          |         |

### Risk Matrix

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| A migration is missed on a route. | Medium | High | User would receive an error when trying to interact with our rule routes if passing back wrong format for investigation_fields |

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.11"]} ONMERGE-->

yctercero committed:
adding migrations at http level

yctercero committed:
added FTR tests for legacy investigation field checks on CRUD routes

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into support_investigation_fields

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/create_rules_stream_from_ndjson.ts:
```diff
@@ -103,6 +104,25 @@ export const sortImports = (): Transform => {
   );
 };
 
+export const migrateRules = (): Transform => {
```
Want to support use case of users exporting their rules in 8.10 and importing in 8.11+ without them needing to manually update this field.

yctercero committed:
removing unnecessary cleanup changes for this PR

yctercero committed:
fixing type check error

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into support_investigation_fields

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

e40pud reviewed (approved):
LGTM!

vitaliidm commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/update_rules_bulk.ts:
```diff
@@ -45,6 +45,7 @@ export default ({ getService }: FtrProviderContext) => {
   const log = getService('log');
   const es = getService('es');
+  const esArchiver = getService('esArchiver');
 
```
can we add test, that not migrated rule continues to execute?

This can happen if rule has investigation fields and is enabled, but not edited by user

yctercero replied:
Manually confirmed rules are now running with updates to code. Thanks for catching this!

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/read_rules.ts:
```diff
@@ -15,6 +15,7 @@ import type {
 import { withSecuritySpan } from '../../../../../utils/with_security_span';
 import type { RuleParams } from '../../../rule_schema';
 import { isAlertType } from '../../../rule_schema';
```
have you had a chance to test bulk edit action? Does it migrate investigation filed for free? Or some work is needed there?

vitaliidm replied:
I can see test for bulk_update, which is deprecated in favour of bulk_actions(bulk_edit - add/removed tags/actions,schedule, etc)

Tests for that one can be found in `x-pack/test/detection_engine_api_integration/security_and_spaces/group10/perform_bulk_action.ts`
I am interested in that one, because it fetches rules within alert framework for editing, so rules are not read through find API, which is covered.

yctercero replied:
Thanks for pointing this out. I'll add tests there now.

marshallmain commented:
I think we still need to handle the case where existing rules have `string[]` data in `investigationFields`, don't we? The internal server-side rule schema has `investigationFields` defined as an object (https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts#L102), but the data could be an array in which case we'll still see rule failures like the ones in https://github.com/elastic/kibana/issues/168682 (`params invalid: Invalid value "[]" supplied to "investigationFields"`).

My suggestion would be to update the server side type to allow both the new object type and the old array type. The API schema changes in this PR make sense to me - migrate the field structure on read and also migrate the structure on rule import so rules from 8.10 can be imported into later Kibana versions. This way our API is clean going forward and only supports the object format.

yctercero committed:
updated server side type (as opposed to api type) to reflect support for both types per Marshalls feedback

yctercero committed:
.

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into support_investigation_fields

spong reviewed (approved):
I don't have anything to add outside of @vitaliidm and @marshallmain's comments, so approving from the rules side 👍

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/create_rules_stream_from_ndjson.ts:
```diff
@@ -103,6 +104,25 @@ export const sortImports = (): Transform => {
   );
 };
 
+export const migrateRules = (): Transform => {
```
Good point - updated!!

yctercero committed:
updating per feedback

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/utils/utils.ts:
```diff
@@ -380,3 +381,37 @@ export const convertAlertSuppressionToSnake = (
       }
     : undefined;
+
+export const migrateRuleLegacyInvestigationFields = (rule: RuleAlertType): RuleAlertType => {
```
Updated!

yctercero committed:
fix test type

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into support_investigation_fields

yctercero commented:
> I think we still need to handle the case where existing rules have `string[]` data in `investigationFields`, don't we? The internal server-side rule schema has `investigationFields` defined as an object (https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts#L102), but the data could be an array in which case we'll still see rule failures like the ones in #168682 (`params invalid: Invalid value "[]" supplied to "investigationFields"`).
> 
> My suggestion would be to update the server side type to allow both the new object type and the old array type. The API schema changes in this PR make sense to me - migrate the field structure on read and also migrate the structure on rule import so rules from 8.10 can be imported into later Kibana versions. This way our API is clean going forward and only supports the object format.

NOTE: Zoomed with Marshall. The SO rule schema will continue to support both types so that we do not get rule failure errors on execution, but the API will only support the intended object type and data will be migrated at the API level. Import is the only caveat, where we will migrate prior to validation in order to allow users from 8.10 to import rules without needing to manually update anything.

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into support_investigation_fields

yctercero committed:
updates needed for bulk duplicate and edit actions

yctercero committed:
Merge branch 'support_investigation_fields' of github.com:yctercero/kibana into support_investigation_fields

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts:
```diff
@@ -47,13 +48,17 @@ export const bulkEditRules = async ({
       await validateBulkEditRule({
         mlAuthz,
-        ruleType: ruleParams.type,
+        ruleType: migratedRuleParams.type,
```
I'm unsure about this change. Without it, doing a bulk action (edit) on a rule with the legacy investigation field fails on params validation. However, the registred params schema defined [here](https://github.com/elastic/kibana/pull/169061/files#diff-c6c196ee2805bef284eea67bc43f94a7fb9f3eeaa0dee2609df3617d2ccbf838R102) includes both the old and new versions, so I'm not sure why this change would be needed.

If someone could confirm my logic here before I merge, would be much appreciated. 

cc @banderror @vitaliidm @marshallmain @spong (pinging you Garrett since you were assigned to this PR).

yctercero replied:
The error I get without this code here is `params invalid: ...` - complains of an array being passed for investigation fields. I traced the error to the bulkEdit rules client logic [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/application/rule/methods/bulk_edit/bulk_edit_rules.ts#L526). You can see that it validates the params against `ruleType.validate.params` - `ruleType` being the options passed when registering a rule type.

When we register the query rule type [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/plugin.ts#L290) - we use this [schema](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts#L92) in the [validation logic](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/query/create_query_alert_type.ts#L41).

This same schema used for validation is what is updated in this PR to allow for the investigation field to be both the old and new types.

maximpn replied:
As far as I see the problem is in [this line](https://github.com/elastic/kibana/pull/169061/files#diff-afcaec4a7447adf94399594770f629a933be7d885a2828dfc1aa10b1acf6393fR298). NonEmptyStringArray expects a non empty string of values separated by commas. If you replace this line with
export const LegacyInvestigationFields = NonEmptyArray(NonEmptyString);
you can roll back the changes in x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts and bulk actions will work.

yctercero replied:
I was able to remove this logic after addressing the issue @maximpn spotted in the type definition.

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into support_investigation_fields

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.ts:
```diff
@@ -286,3 +292,21 @@ export const InvestigationFields = t.exact(
+
+// 8.10.x is mapped as an array of strings
+export type LegacyInvestigationFields = t.TypeOf<typeof LegacyInvestigationFields>;
+export const LegacyInvestigationFields = NonEmptyStringArray;
```
As @maximpn found out, this is an incorrect type. The schema should be either of these:

```ts
export const LegacyInvestigationFields = NonEmptyArray(NonEmptyString);
export const LegacyInvestigationFields = t.array(NonEmptyString);
export const LegacyInvestigationFields = NonEmptyArray(t.string);
export const LegacyInvestigationFields = t.array(t.string);
```

Probably the latter one will give the most flexibility. We would accept any kinds of arrays: empty and non-empty, also containing empty strings potentially, and will have to handle these edge cases in the migration function on the server side.

banderror reviewed (changes requested):
Posting some comments after a quick review. I didn't have time to do a proper review and didn't test the changes. I can do this later today if needed.

I think we need to fix a couple of bugs in the code before merging this PR.

yctercero commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.ts:
```diff
@@ -286,3 +292,21 @@ export const InvestigationFields = t.exact(
   })
 );
+
+// 8.10.x is mapped as an array of strings
+export type LegacyInvestigationFields = t.TypeOf<typeof LegacyInvestigationFields>;
+export const LegacyInvestigationFields = NonEmptyStringArray;
+
+/*
+ * In ESS 8.10.x "investigation_fields" are mapped as string[].
+ * For 8.11+ logic is added on read in our endpoints to migrate
+ * the data over to it's intended type of { field_names: string[] }.
+ * The SO rule type will continue to support both types until we deprecate,
+ * but APIs will only support intended object format.
+ * See PR 169061
+ */
+export type InvestigationFieldsCombined = t.TypeOf<typeof InvestigationFieldsCombined>;
+export const InvestigationFieldsCombined = t.union([
+  InvestigationFields,
+  LegacyInvestigationFields,
+]);
```
Moving over as suggested and implemented the type change mentioned by @maximpn (will push up after cleanup).

banderror commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/perform_bulk_action.ts:
```diff
@@ -2997,5 +2997,223 @@ export default ({ getService }: FtrProviderContext): void => {
+        const ruleIdWithLegacyInvestigationFieldEmptyArray = '2297be91-894c-4831-830f-b424a0ec5678';
+        const ruleIdWithInvestigationField = '2297be91-894c-4831-830f-b424a0ec9102';
+
+        const { body } = await postBulkAction()
+          .send({ query: '', action: BulkActionType.delete })
+          .expect(200);
+
+        expect(body.attributes.summary).to.eql({ failed: 0, skipped: 0, succeeded: 3, total: 3 });
+
+        // Check that the deleted rule is returned with the response
+        const names = body.attributes.results.deleted.map(
+          (returnedRule: RuleResponse) => returnedRule.name
+        );
+        expect(names.includes('Test investigation fields')).to.eql(true);
+        expect(names.includes('Test investigation fields empty array')).to.eql(true);
+        expect(names.includes('Test investigation fields object')).to.eql(true);
+
+        // Check that the updates have been persisted
+        await fetchRule(ruleIdWithLegacyInvestigationField).expect(404);
+        await fetchRule(ruleIdWithLegacyInvestigationFieldEmptyArray).expect(404);
+        await fetchRule(ruleIdWithInvestigationField).expect(404);
```
This test checks that the overall bulk deletion logic works, but doesn't check anything related to the investigation fields specifically. I think we should delete it or add any specific checks for the field.

Same for other tests:

- "should enable rules with legacy investigation fields"
- "should disable rules with legacy investigation fields"

yctercero replied:
Updated tests to be more specific about what they are testing and if SO migration is inteded.

banderror commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/perform_bulk_action.ts:
```diff
@@ -2997,5 +2997,223 @@ export default ({ getService }: FtrProviderContext): void => {
+        expect(ruleBody3.enabled).to.eql(false);
+      });
+
+      it('should duplicate rules with legacy investigation fields', async () => {
```
This one also doesn't check anything for the field specifically, although it could assert that in the duplicates the field is migrated.

yctercero replied:
Updated tests to be more specific about what they are testing and if SO migration is inteded.

banderror commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/perform_bulk_action.ts:
```diff
@@ -2997,5 +2997,223 @@ export default ({ getService }: FtrProviderContext): void => {
+        expect(rulesResponse.total).to.eql(6);
+      });
+
+      it('should edit rules with legacy investigation fields', async () => {
```
```suggestion
      it('should migrate legacy investigation fields on bulk editing', async () => {
```

banderror reviewed (commented):
Did a more thorough review and left a few more comments, most of them are minor. I didn't test the PR locally.

yctercero committed:
addressed additional feedback and increased test coverage

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into support_investigation_fields

yctercero committed:
Merge branch 'support_investigation_fields' of github.com:yctercero/kibana into support_investigation_fields

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

yctercero commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/delete_rules.ts:
```diff
@@ -239,5 +240,57 @@ export default ({ getService }: FtrProviderContext): void => {
       });
     });
+
+    describe('legacy investigation fields', () => {
+      before(async () => {
+        await esArchiver.load(
+          'x-pack/test/functional/es_archives/security_solution/legacy_investigation_fields'
+        );
+        await createSignalsIndex(supertest, log);
+      });
```
I think that could work, if it's ok, I'll follow up on this clean up.

yctercero replied:
I ended up just doing it here, so no more esarchiver used.

yctercero committed:
removed usage of esarchiver in favor of API

yctercero commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/patch_rules.ts:
```diff
@@ -671,5 +672,77 @@ export default ({ getService }: FtrProviderContext) => {
+        );
+      });
+
+      it('migrates investigation fields when array exists', async () => {
```
Already done 😉

yctercero committed:
jest cleanup for use case that does not match type

yctercero committed:
Merge branch 'support_investigation_fields' of github.com:yctercero/kibana into support_investigation_fields

yctercero commented:
@banderror @maximpn @vitaliidm @e40pud 

Thanks for all the feedback here. I updated the tests so that they now explicitly test the SOs and response where needed. So it should now be clear from the test if the expected behavior is for the SO to be migrated or if it's just expected that the response will be transformed. I also added a few more tests for some routes that had already been migrated into the new folder for serverless + ess FTR tests. One other test added was a test in of the rule execution - checking that a rule  with the legacy investigation field that gets enabled (no migration on bulk enable), successfully runs and generates alerts.

yctercero commented on x-pack/test/detection_engine_api_integration/utils/create_rule_saved_object.ts:
```diff
@@ -0,0 +1,35 @@
+  InternalRuleCreate,
+} from '@kbn/security-solution-plugin/server/lib/detection_engine/rule_schema';
+
+/**
```
I thought about that, but was also trying to avoid hitting any possible circular dependency issues. Since the work to move over these tests all into `security_solution_api_integration` is a work in progress and should be resolved in the next weeks, I thought it ok to play it safe here and just duplicate.

yctercero commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/perform_bulk_action.ts:
```diff
@@ -2997,5 +3003,402 @@ export default ({ getService }: FtrProviderContext): void => {
+         * Confirm type on SO so that it's clear in the tests whether it's expected that
+         * the SO itself is migrated to the inteded object type, or if the transformation is
+         * happening just on the response. In this case, change should not include a migration on SO.
+         */
```
I'm updating the description of the ticket with a table that specifies the expected migration/transform behavior for all the routes.

It's not migrating right now, but maybe it makes sense to since PUT/PATCH rule endpoints do. @banderror any thoughts here?

marshallmain replied:
IMO it would be simpler to avoid having routes migrate the internal data structure in most cases. I don't think there's a big advantage to a migrated SO vs a legacy SO since I wouldn't expect the new structure to be more performant and we need to handle both structures in the server side code anyway since we can't guarantee that all SOs will ever be migrated.

With that line of thinking, `import` could be the only special case that does migrate the structure on the way in to allow importing rules from 8.10. Every other route would convert the data before returning it, but would not have to migrate the stored data format.

banderror replied:
I'm fine with @marshallmain's suggestion because it simplifies the code a lot.

yctercero committed:
some cleanup of comments and test descriptions

marshallmain commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts:
```diff
@@ -442,7 +446,9 @@ export const convertPatchAPIToInternalSchema = (
       ruleId: existingParams.ruleId,
       falsePositives: nextParams.false_positives ?? existingParams.falsePositives,
-      investigationFields: nextParams.investigation_fields ?? existingParams.investigationFields,
+      investigationFields: migrateLegacyInvestigationFields(
+        nextParams.investigation_fields ?? existingParams.investigationFields
+      ),
```
Related to https://github.com/elastic/kibana/pull/169061#discussion_r1372377694, I'd consider omitting the migration logic here to make it easier to reason about which routes will migrate data as it comes into the server. If we remove the migration here, I believe the import route will be the only route to modify `investigation_fields` on write. Every other route will simply pass through whatever it receives from the API, which now in 8.11 will always be the new object-based type. PATCH based routes would leave `investigation_fields` un-modified if the field is not specified in the request.
```suggestion
      investigationFields: nextParams.investigation_fields ?? existingParams.investigationFields,
```

yctercero replied:
Updated so now we are only migrating the response and in rule import. PATCH will not migrate unless specified in request.

marshallmain commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_rule_with_fallback.ts:
```diff
@@ -114,11 +116,50 @@ export const useRuleWithFallback = (ruleId: string): UseRuleWithFallback => {
+ * @param investigationFields InvestigationFieldsCombined | undefined
+ * @returns InvestigationFields | undefined
+ */
+export const migrateLegacyInvestigationFields = (
```
nit: There are 3 different functions called `migrateLegacyInvestigationFields`, can we differentiate them with more descriptive names?

marshallmain reviewed (approved):
My main concern is addressed since the internal SO schema now represents both types and the API schema allows only the new object-based type. I left 2 more comments on potential improvements.

yctercero committed:
updated per feedback to have migration only occur on response transform and import

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into support_investigation_fields

yctercero committed:
cleanup

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/read_rules.ts:
```diff
@@ -47,7 +48,7 @@ export const readRules = async ({
             return restOfRule;
           }
-          return rule;
+          return migrateRuleLegacyInvestigationFields(rule);
```
Before making the change Marshall suggested of continuing to support both types in the SO, this was needed. But now it no longer is and was removed.

yctercero committed:
cleanup

vitaliidm reviewed (approved):
Thanks for addressing comments, adding so extensive test coverage.
Special thanks for API table, that allows at one glimpse to understand scope of changes.

The recent shift for not migrating SO, not something what I would expect in scenarios like this(we won't be able to use only one internal structure over the time, but will have to support both indefinitely)
But it does it job and ensures we cover both scenarios

banderror reviewed (approved):
The updated integration tests are extremely comprehensive, clear, and just amazing! I love the explicit SO checks and the table you added in the description of the PR (nit: I assume the PUT and PATCH endpoints should be marked as Legacy blocked in request ✅). Thank you for addressing my comments @yctercero, I'm leaving a bunch of minor comments that don't block this PR.

yctercero commented:
Thank you so much @banderror @vitaliidm @marshallmain for the thorough review!

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.11|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 169061
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

yctercero commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.11|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/169957">](https://github.com/elastic/kibana/pull/169957)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->
